src/libsyntax/parse/parser.rs

```diff
@@ -5049,20 +5096,25 @@ impl<'a> Parser<'a> {
                     bounds,
                     kind: ast::GenericParamKind::Lifetime,
                 });
-                if seen_ty_param {
-                    self.span_err(self.prev_span,
-                        "lifetime parameters must be declared prior to type parameters");
-                }
+                enforce_param_order(&self, &mut max_param, ParamKindOrd::Lifetime);
+            } else if self.check_const_param() {
+                // Parse const parameter.
+                params.push(self.parse_const_param(attrs)?);
+                enforce_param_order(&self, &mut max_param, ParamKindOrd::Const);
             } else if self.check_ident() {
                 // Parse type parameter.
                 params.push(self.parse_ty_param(attrs)?);
-                seen_ty_param = true;
+                enforce_param_order(&self, &mut max_param, ParamKindOrd::Type);
             } else {
                 // Check for trailing attributes and stop parsing.
                 if !attrs.is_empty() {
-                    let param_kind = if seen_ty_param { "type" } else { "lifetime" };
-                    self.span_err(attrs[0].span,
-                        &format!("trailing attribute after {} parameters", param_kind));
+                    if let Some(max_param) = max_param {
+                        self.span_err(attrs[0].span,
+                            &format!("trailing attribute after {} parameters", max_param));
+                    } else {
+                        self.span_err(attrs[0].span,
+                            "leading attribute before generic parameters");
```

Comment:
Leading attributes before generic parameters are allowed.
This case means something like `<#[attr]>`, i.e. no parameters at all.
I'd keep the old wording (maybe with replacing "lifetime parameters" with "generic parameters").